proposals/roles.md

```diff
@@ -0,0 +1,155 @@
+﻿# Roles
+
+This proposal is divided into three parts, all relating to extending existing types:
+A. static extensions
+B. roles and extensions with members
+C. roles and extensions that implement interfaces
+
+The syntax for a roles is as follows:
+
+```antr
+type_declaration
+	: role_declaration // add
+	| extension_declaration // add
+	| ...
+	;
+
+role_declaration
+    : modifier* 'role' identifier type_parameter_list? role_underlying_type type_parameter_constraints_clause* role_body
+    ;
+
+extension_declaration
+    : modifier* 'extension' identifier type_parameter_list? role_underlying_type type_parameter_constraints_clause* role_body
+    ;
+
+role_underlying_type
+    : ':' type
+    ;
+
+role_body
+    : '{' role_member_declaration* '}'
+    ;
+
+role_member_declaration
+    : constant_declaration
+    | field_declaration
+    | method_declaration
+    | property_declaration
+    | event_declaration
+    | indexer_declaration
+    | operator_declaration
+    | type_declaration
```

Comment:
allowing type-decls in roles is def interesting (but seems like it could be useful).  What is teh fully qualified name of that type-decl though?  

are roles also allowed to be defined in roles?

Reply:
> What is teh fully qualified name of that [nested] type-decl though?

Just like any other nested type-decl. You can refer to it with `Extension.NestedType`.

> are roles also allowed to be defined in roles?

See below, "In this first subset of the feature, the syntax is restricted to non-nested extension_declaration"

Reply:
> You can refer to it with Extension.NestedType.

Nifty.  Is that in metadata as well?

Reply:
TBD. There's an open question of what name we use to emit the extension type.